Microsoft.Research/Contracts/System.Windows.Forms/System.Windows.Forms.DataGridViewCellCollection.cs

```diff
@@ -0,0 +1,164 @@
+// 
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+// 
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+using System;
+using System.Diagnostics.Contracts;
+using System.Collections;
+using System.ComponentModel;
+
+namespace System.Windows.Forms
+{
+    /// <summary>
+    /// Represents a collection of cells in a <see cref="T:System.Windows.Forms.DataGridViewRow"/>.
+    /// </summary>
+    
+    public class DataGridViewCellCollection : BaseCollection // IList, ICollection, IEnumerable
+    {
+    
+        /// <summary>
+        /// Gets an <see cref="T:System.Collections.ArrayList"/> containing <see cref="T:System.Windows.Forms.DataGridViewCellCollection"/> objects.
+        /// </summary>
+        /// 
+        /// <returns>
+        /// <see cref="T:System.Collections.ArrayList"/>.
+        /// </returns>
+        // protected override ArrayList List
+        
+        /// <summary>
+        /// Gets or sets the cell at the provided index location. In C#, this property is the indexer for the <see cref="T:System.Windows.Forms.DataGridViewCellCollection"/> class.
+        /// </summary>
+        /// 
+        /// <returns>
+        /// The <see cref="T:System.Windows.Forms.DataGridViewCell"/> stored at the given index.
+        /// </returns>
+        /// <param name="index">The zero-based index of the cell to get or set.</param><exception cref="T:System.ArgumentNullException">The specified value when setting this property is null.</exception><exception cref="T:System.InvalidOperationException">The specified cell when setting this property already belongs to a <see cref="T:System.Windows.Forms.DataGridView"/> control.-or-The specified cell when setting this property already belongs to a <see cref="T:System.Windows.Forms.DataGridViewRow"/>.</exception><exception cref="T:System.ArgumentOutOfRangeException"><paramref name="index"/> is less than 0.-or-<paramref name="index"/> is equal to or greater than the number of cells in the collection.</exception>
+        public DataGridViewCell this[int index]
+        {
+            get
+            {
+                Contract.Requires(index >= 0 && index <= this.Count);
+                return default(DataGridViewCell);
+            }
+            set
+            {
+                Contract.Requires(value != null);
```

Comment:
We can use stronger precondition (or add few more):

`Contract.Requires(value.DataGridView == null);`
`Contract.Requires(value.OwningRow == null);`